src/dune_pkg/repository_id.ml

```diff
@@ -0,0 +1,83 @@
+open Stdune
+module Process = Dune_engine.Process
+module Display = Dune_engine.Display
+
+type t =
+  | Git_hash of string
+  | Opam_stamp of string
+
+let equal a b =
+  match a, b with
+  | Git_hash a, Git_hash b | Opam_stamp a, Opam_stamp b -> String.equal a b
+  | _, _ -> false
+;;
+
+let encode : t -> Dune_sexp.t = function
+  | Git_hash commitish ->
+    List
+      [ Dune_lang.atom_or_quoted_string "Git_hash"
+      ; Dune_lang.atom_or_quoted_string commitish
+      ]
+  | Opam_stamp stamp ->
+    List
+      [ Dune_lang.atom_or_quoted_string "Opam_stamp"
+      ; Dune_lang.atom_or_quoted_string stamp
+      ]
+;;
+
+let decode =
+  let open Dune_sexp.Decoder in
+  let+ constr, stamp = pair string string in
+  match constr with
+  | "Git_hash" -> Git_hash stamp
+  | "Opam_stamp" -> Opam_stamp stamp
+  | _ -> failwith "TODO: parsing failure"
+;;
+
+let to_dyn = function
+  | Git_hash commitish -> Dyn.Variant ("Git_hash", [ String commitish ])
+  | Opam_stamp stamp -> Dyn.Variant ("Opam_stamp", [ String stamp ])
+;;
+
+let repo_id_of_stamp dir =
+  let local = Path.Local.of_string "repo" in
+  let repo_file = Path.append_local dir local in
+  match Path.exists repo_file with
+  | false ->
+    (* no repo file, we can't read this, bail *)
```

Comment:
`dune pkg lock` will now behave in subtly different ways depending on whether `git` is installed. That's fine but I think we should print a warning in this case to help clarify the situation.

Reply:
I'm not super happy about this either (though I would assume git has a high installation base and be able to fetch a git repo you needed to have git already). The alternative could be to determine the revision from the `.git` directory manually. This isn't all that difficult given the needs of this PR (parse `.git/HEAD`, look up the referenced ref, read that file for the hash) but I am not familiar with how much of the `.git` directory is a "public" API and how much are implementation details.

Let me know what you think. Removing the `git` binary avoids all the path lookups but makes me (slightly) concerned about the future-proofness of reading `.git` directory (or we could use `ocaml-git` but for such a simple task that would be extreme overkill).

Reply:
Since the only difference in behaviour between systems with git installed and those without it will be that the latter will lack the repo id in package metadata I think it's ok to just omit the field when git isn't installed. There won't be a functional difference as this field will just be read by humans to help understand how their lockdir was generated and as you said most users will already have git installed. In the case that git isn't installed I much prefer printing a warning indicating that the metadata will lack the repo id because git isn't installed rather than having dune try to parse the .git directory as it's simpler to implement for us and I think less surprising behaviour for users.

Reply:
I've solved it by reimplementing that part parsing the `.git` data, so it works whether you have `git` installed or not, which also solves the error handling if the repo can't be parsed by `git`.

This part of git is fairly simple so I am not too worried about it breaking.